Implemented skip violations

torinaki opened the pull request:
Initial implementation of #199

## Scope
Implement possibility to skip class dependency violation in `depfile.yml`. There will be new configuration section `skip_violations` (maybe `skipped_violations` better?). The developer can configure it in the following way:
```yaml
skip_violations:
  Library\LibClass:
    - Core\CoreClass
    - Core\AnotherClass
```
`skip_violations` section contains associative array where key (`Library\LibClass`) is the name of dependant class and values (`Core\CoreClass`, `Core\AnotherClass`) are dependency classes.

## Future scope
There might be implemented possibility for PHP functions dependency as well. In this case, we can specify functions names instead of class names. PHP allows to create functions with names which can be equal to class names in the same namespace:
```php
<?php

namespace Foo;

class bar {
    function __invoke()
    {
        echo __CLASS__ . PHP_EOL;
    }
}

function bar() {
    echo __FUNCTION__ . PHP_EOL;
}

bar();
(new bar())();
```
output:
```text
Foo\bar
Foo\bar
```
So function names should be identified differently from clasess. We can sufix each function name with `()`:
```yaml
skip_violations:
  Library\LibClass:
    - Foo\bar()
    - Foo\bar
```

## TODO
- [x] Implement class violations skipping possibility
- [x] Unit tests
- [x] Update documentation

timglabisch commented:
thanks for trying to make deptrac better, 

i am unsure about this PR. I would agree that we need some way to skip violations but there are a bunch of issues with this approach.

when i wrote deptrac i had the idea that a dependency is something abstract. in fact  we (mostly) use deptrac on the "class"-level but if we think further dependencies has nothing to do with classes. 

for example we could write a npath collector that collects every function that is "too complex". than you could write a rule to prevent some parts of your code to prevent on too complex code. You would end up with a dependency between (may) 2 functions, not classes.
`skip_violations` is only build on `class` level.

atm. i am unsure about how to solve this issue but i guess the approach is not generic enough.

torinaki commented:
@timglabisch agree with you. 
Talking about classes and functions. I can change the configuration format to support both functions and classes to avoid configuration BC breaks. Or it might be improved in future releases until `\SensioLabs\Deptrac\DependencyResult\DependencyInterface` supports only classes.

When you integrate DepTrac in CI/CD with large code base (our case) all dependencies cannot be easily fixed, because it would take a lot of developer-hour. IMO DepTrac definitely should support violation skip possibility otherwise integration process might not be so smooth.

torinaki commented:
@timglabisch as you can see in #199 we have implemented a workaround using existing functionality, but not the best option. We would like to see jUnit support as well to see and control the amount of existing skipped violations. Also, an approach described in #199 provides a possibility to skip some class completely from an analysis, but in the scope of this PR I have implemented a possibility to skip exact violation which minimizes probability to introduce new violations.

torinaki commented:
@timglabisch any ideas? Should I continue work on PR or you will just close it?

smoench commented:
I think this would be good start having a skipping mechanism. If there is somebody having the use case skipping functions/npath then they/we should start thinking about abstracting the mechanism.

@torinaki 👍 feel free to continue

smoench commented on README.md:
```diff
@@ -319,6 +319,40 @@ Found 2 Violations
 
+### Skip violations
+
+DepTrac integration into existing CI/CD pipeline might be difficult because of existing dependency violations in the code.
```
`Deptrac`

torinaki replied:
Fixed

torinaki commented:
@smoench , @timglabisch it seems that PR is ready for review.

DavidBadura commented on src/RulesetEngine.php:
```diff
@@ -45,4 +46,23 @@ public function getViolations(Result $dependencyResult, ClassNameLayerResolverIn
     }
+
+    /**
+     * Get skipped violations
```
This comment brings no added value, too.

torinaki replied:
fixed

smoench commented:
@torinaki Would be great if you could rebase on actual master. I extended the travis ci config to enforce static analysis. As soon as the pipeline gets green I will merge your PR :)

torinaki commented on src/Configuration/ConfigurationSkippedViolation.php:
```diff
@@ -0,0 +1,36 @@
+namespace SensioLabs\Deptrac\Configuration;
+
+/**
+ * Class Configuration skipped violation
```
fixed

torinaki commented on src/Configuration/ConfigurationSkippedViolation.php:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    /**
+     * @return bool
```
fixed

smoench commented:
@torinaki It seems you have just a few coding style issues. They could be easily fixed with [php-cs-fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer). Just download the latest phar-file and run following command `php php-cs-fixer.phar fix`. Thanks :)

smoench commented:
Thank you @torinaki
